### ISSUE TITLE
CA1812 incorrect message about static class not being instantiated

### ISSUE THREAD

**Issue by RobSiklos** [dotnet/roslyn-analyzers#4389]
This is a regression - the problem didn't happen in v3.3.0.

### Analyzer

**Diagnostic ID**: [CA1812](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1812): `Avoid uninstantiated internal classes`

### Analyzer source

**NuGet Package**: [Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

**Version**: v3.3.1 (Latest)

### Steps To Reproduce

Create an internal static class without any public or internal methods.

```csharp
    internal static class Foo
    {
    }
```

### Actual behavior

Warning reported:

> CA1812: Foo is an internal class that is apparently never instantiated. If so, remove the code from the assembly. If this class is intended to contain only static members, make it static (Shared in Visual Basic)

### Additional context

This message doesn't make any sense - the class is already marked as `static`.

**Comment by mavasani** [comment]
Tagging @Evangelink

**Comment by Evangelink** [comment]
Indeed, there is even a test for this case. I see two ways to fix this case:

1. We want to be strict about the instantiation, in which case we stop reporting for this case.

2. We want the rule to be more around "use", in which case we could create another descriptor with a different message (i.e. `Foo is an internal class that is apparently never used. If so, remove the code from the assembly.`).

WDYT?

**Comment by Youssef1313** [comment]
@Evangelink The suggestion second option would be better as a separate rule IMO.

I think the analyzer should just exclude static types. It seems that the value returned from the `IsStaticHolderType` call below is false:

https://github.com/dotnet/roslyn-analyzers/blob/c26b04af4f237e17fc50053339d6641c9df67b29/src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidUninstantiatedInternalClasses.cs#L349-L352

Not sure why it is?

Note: the containing method for the above check is public, and has a misleading name:

https://github.com/dotnet/roslyn-analyzers/blob/c26b04af4f237e17fc50053339d6641c9df67b29/src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidUninstantiatedInternalClasses.cs#L285

should be private and named `IsOkToBeUninstantiated`?

I'm also unsure why we care whether the type is a "static holder" or not? shouldn't we only care whether it is "static" or not?

**Comment by Evangelink** [comment]
@Youssef1313 I am not sure what is supposed to be the exact scope of this rule so I'd rather wait for @mavasani feedback before going for a fix in one direction or another.

**Comment by Evangelink** [comment]
Ping @mavasani

**Comment by mavasani** [comment]
[Rule description](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1812#rule-description) of this rule explicitly mentions it is not reported for non-static types. I think its probably best to report unused static types with a completely different rule ID. IMO it will be easier to report it in the same analyzer as `CA1812` by just adding a new descriptor to the analyzer and then choosing the correct descriptor for `ReportDiagnostic` call.

**Comment by Youssef1313** [comment]
@mavasani I'm not sure which part of the rule description mentions it's not reported for non-static types?
The rule makes sense for non-static types.

The rule description states that "Types that cannot be instantiated" are not examined by the rule (i.e. not reported).

**Comment by mavasani** [comment]
@Youssef1313 sorry I had a typo in my comment - fixed it now. I meant let's restrict `CA1812` to non-static types.

**Comment by Youssef1313** [comment]
@mavasani I attempted to do that in my PR, but that will cause the issue #1878 to be back. Should I use `IsStaticHolderType` only for VB?

**Comment by mavasani** [comment]
That sounds reasonable to me. @Evangelink your thoughts?

**Comment by Youssef1313** [comment]
Note: that will require doc update on the following sentence:

> Types that cannot be instantiated and that only define static (Shared in Visual Basic) methods.

It will need to be updated to:

> Types that cannot be instantiated and that only define Shared (in Visual Basic) methods.

**Comment by mavasani** [comment]
@Youssef1313 - I think there is still some confusion here:

1. `CA1812` should stick to the current documentation, i.e. do not report any `CA1812` diagnostics for both C# and VB static types that only define static methods. Additionally, do not regress https://github.com/dotnet/roslyn-analyzers/issues/1878 by skipping VB non-static types that define a private constructor (static holder type).
2. Introduce a new diagnostic ID to report unused static types for C# and VB.

**Comment by Youssef1313** [comment]
>  i.e. do not report any CA1812 diagnostics for both C# and VB static types that only define static methods.

What I understand from the documentation is that the rule doesn't report any diagnostic for types that only define static methods (whether the types themselves are declared static or not).

So I think #4395 matches the current documentation.
